### ISSUE TITLE
[plan] MobaMatchState keeps freed combatants in its rosters: null guards do not catch a freed instance

### ISSUE THREAD

**Issue by stardustsuperwizard** [stardustsuperwizard/mikeys_game_bones-rules-moba#371]
## Summary

`MobaMatchState` guards its roster iteration with `combatant != null`. That
check does not catch a combatant whose node has been freed: in GDScript a freed
Object is not equal to `null`, and assigning one to a typed loop variable
throws outright.

Every roster walk therefore errors once any registered combatant is freed —
which is what a despawn, a scene teardown, or an actor being removed between
rounds does.

`tests/death_respawn_camera_test.gd` fails on `main` because of it.

## Expected Behavior

A combatant freed while the match is running is treated as absent — the same
way a `null` entry already is ("A null entry counts as nobody"). Round and
match lifecycle continue without error.

## Actual Behavior

Four script errors per tick and a failed round transition:

```
SCRIPT ERROR: Trying to assign invalid previously freed instance.
          GDScript backtrace (most recent call first):
              [0] _living_count (res://rules/core/moba_match_state.gd:162)
              [1] tick (res://rules/core/moba_match_state.gd:143)
              [2] _physics_process (res://scripts/match_controller.gd:41)
SCRIPT ERROR: Trying to assign invalid previously freed instance.
          GDScript backtrace (most recent call first):
              [0] start_round (res://rules/core/moba_match_state.gd:125)
              [1] tick (res://rules/core/moba_match_state.gd:153)
              [2] _physics_process (res://scripts/match_controller.gd:41)
FAIL death: expected state DEAD, got 0
```

## Reproduction Steps

```bash
godot --headless --path . --script tests/death_respawn_camera_test.gd
```

Observed: `FAIL death: expected state DEAD, got 0`, preceded by the errors
above. Expected: the suite passes.

## Reproduction Rate

Always, on `main` @ `a9673e6`.

## Environment

- Godot version: 4.7.1-stable
- Branch/commit: `main` @ `a9673e6` (`[351] Inspect another player's build from
  the lobby (#367)`)
- Platform: platform-independent

## Evidence

Both roster walks use the same insufficient guard.

`rules/core/moba_match_state.gd:162` — the error line:

```gdscript
## Living combatants on one team. A null entry counts as nobody.
func _living_count(team_id: int) -> int:
	var living := 0
	for combatant: MobaCombatant in _rosters[team_id]:
		if combatant != null and not combatant.is_dead():
			living += 1
	return living
```

`rules/core/moba_match_state.gd:121-126` — the same shape in `start_round()`:

```gdscript
func start_round() -> void:
	for roster: Array in _rosters:
		for combatant: MobaCombatant in roster:
			if combatant != null:
				combatant.reset_for_round()
```

The throw happens at the loop variable's typed assignment, *before* the body's
`!= null` runs, so moving or strengthening the check inside the body is not
sufficient on its own.

The docstring "A null entry counts as nobody" states the intended contract
correctly. The implementation just does not cover the freed case, which is a
different thing from `null` and the one that actually occurs.

## Suspected Scope (Optional)

`rules/core/moba_match_state.gd`, both roster walks — `is_instance_valid()`
rather than (or in addition to) `!= null`, applied so the freed entry is
skipped before it reaches a typed assignment.

Worth deciding rather than assuming: whether rosters should also be pruned of
freed entries at some point in the lifecycle, or whether skipping on read is
the whole fix. Registration/deregistration is the other half of this — a
combatant that is despawned deliberately arguably ought to deregister rather
than be left to be detected as freed. That is a design call for the planner,
not something to settle inside the correction.

## Constraints

- `rules/` keeps its one-way dependency arrow. The fix stays inside
  `rules/core/`, and must not reach into `scripts/`.
- Fix the underlying defect rather than suppressing the symptom. Silencing the
  error without making the round transition correct leaves
  `death_respawn_camera_test.gd` failing.
- Do not weaken `tests/death_respawn_camera_test.gd` to make it pass.

## Acceptance Criteria

- [ ] A combatant freed mid-match is skipped by `_living_count()` and
      `start_round()` without a script error.
- [ ] Round transitions resolve correctly when a roster holds a freed entry —
      `tests/death_respawn_camera_test.gd` passes.
- [ ] A `rules/` test covers the freed-entry case directly, so this cannot
      regress silently once #370 makes the `tests/` suites gate a merge.
- [ ] `.github/scripts/validate-godot.sh` still passes.

## Out of Scope

- Wiring the `tests/` suites into CI — #370.
- Redesigning round or match lifecycle. This is a correction to roster
  iteration.

## Notes

Found while implementing #352 / #369, by running the standalone `tests/` suites
that CI does not run (#370). Not caused by that change: the failure reproduces
identically with `scripts/world_manager.gd` reverted to `main`.

Filed rather than fixed in that PR, at the requesting user's explicit
instruction to file discovered out-of-scope work.
